### PR TITLE
Add unit tests for ISIS_Powder common routines

### LINKED ISSUE #19113: ISIS Powder - Unit tests for common routines
Now that GEM porting has been mostly completed (#19112) with the majority of large changes this would be a good time to start writing unit tests for all the common methods.

At a later date we can address instrument specific unit tests

### PR DESCRIPTION
**Pull request by DavidFair**
~~**Must not be merged before #19112** - I would recommend waiting for that PR before taking this one so you can get an accurate diff~~

**Description of work.**
This PR adds unit tests for the common functions in the ISIS_Powder scripts

**To test:**
Check that unit tests pass
Code review

Fixes #19113 .

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [x] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [x] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/Diffraction/isis_powder/routines/InstrumentSettings.py [hunk 1]
```diff
@@ -65,9 +65,9 @@ def update_attributes(self, advanced_config=None, basic_config=None, kwargs=None
             self._parse_attributes(self._adv_config_dict, suppress_warnings=suppress_warnings)
         if advanced_config or basic_config:
             self._parse_attributes(self._basic_conf_dict,
-                                   suppress_warnings=(not bool(basic_config or suppress_warnings)))
+                                   suppress_warnings=(not bool(basic_config) or suppress_warnings))
```
Comment (DimitarTasev) [inline scripts/Diffraction/isis_powder/routines/InstrumentSettings.py]: More of a query than a note, do we have to cast implicitly to `bool`? I think it's done automatically when you put the `not` in front, i.e. there shouldn't be a difference between `not bool(basic_config)` and just `not basic_config`

**Comment by DimitarTasev** [comment]
From a brief look changes look good, all the unit tests pass, I'll have a look again after #19112 is merged and mark for ship then

**Comment by DimitarTasev** [comment]
From a brief look changes look good, all the unit tests pass, I'll have a look again after #19112 is merged and mark for ship then

File: scripts/Diffraction/isis_powder/routines/InstrumentSettings.py (hunk 1 above)
Comment (DavidFair) [inline scripts/Diffraction/isis_powder/routines/InstrumentSettings.py]: Good point Python does this implicitly I'll change that in the current branch I'm working on (to reduce the number of merge conflict in the future)

File: scripts/Diffraction/isis_powder/routines/yaml_parser.py [hunk 2]
```diff
@@ -56,8 +56,9 @@ def _find_dictionary_key(dict_to_search, run_number):
                 generated_runs = common.generate_run_numbers(run_number_string=key)
             except RuntimeError:
                 raise ValueError("Could not parse '" + str(key) + "'\n"
-                                 "This should be a range of runs in this cycle in the mapping file."
-                                 " Please check your indentation if this should be within a cycle.")
+                                 "This should be a range of runs in the mapping file."
+                                 " Please check your indentation if the syntax looks correct.")
```
Comment (DimitarTasev) [inline scripts/Diffraction/isis_powder/routines/yaml_parser.py]: `Please check your indentation if the syntax looks correct.` consider rephrasing to something like
`Please check your indentation and make sure that the syntax is correct.`

File: scripts/Diffraction/isis_powder/routines/common.py [hunk 3]
```diff
@@ -156,7 +157,10 @@ def get_first_run_number(run_number_string):
     :return: The first run for the user input of runs
     """
     run_numbers = generate_run_numbers(run_number_string=run_number_string)
-    return run_numbers[0]
+    if isinstance(run_numbers, list):
```
Comment (DimitarTasev) [inline scripts/Diffraction/isis_powder/routines/common.py]: It seems like `generate_run_numbers` _always_ returns a list, with single run number, or a range

**Review by dtasev — CHANGES_REQUESTED** [review]

File: scripts/Diffraction/isis_powder/routines/common.py (hunk 3 above)
Comment (DimitarTasev) [inline scripts/Diffraction/isis_powder/routines/common.py]: also in `get_monitor_ws` [here](https://github.com/mantidproject/mantid/pull/19163/files#diff-3e56862968e7fc4f56f839cc920d48baL174), I think you could use get_first_run_number

**Comment by DimitarTasev** [comment]
Use `get_first_run_number` in `yaml_parser.py` [here](https://github.com/mantidproject/mantid/blob/ac05bc1459672ae164348155091688aed2209027/scripts/Diffraction/isis_powder/routines/yaml_parser.py#L11) ?

File: scripts/Diffraction/isis_powder/routines/common.py (hunk 3 above)
Comment (DavidFair) [inline scripts/Diffraction/isis_powder/routines/common.py]: Thats correct - it means the caller doesn't have to figure out types on their end (e.g. can just use a for in loop or [0])

Good spot 👍

**Review by dtasev — APPROVED** [review]

**Comment by DimitarTasev** [comment]
:shipit: when OSX build lands on an ISIS node which has yaml installed

**Pull request merged**